KAFKA-13881: Add Clients package description javadocs

gharris1727 opened the pull request:
Signed-off-by: Greg Harris <greg.harris@aiven.io>

There are currently no descriptions of packages in the javadocs, which can be intimidating to new users that look at the documentation and see a wall of raw package names. An example of this can be seen here: https://kafka.apache.org/31/javadoc/

This change adds the following package descriptions based on a best-effort reading of the existing javadocs. Please feel free to correct any factual errors in the proposed descriptions, I am not asserting that these descriptions are currently accurate. This PR only covers the `clients` packages for core kafka.

For clarity of review, I'll summarize the new descriptions in a table. This summary may become stale by accident, so please be sure to view the diff itself or build the changes locally with `./gradlew javadoc`.


Package | Description
--- | ---
org.apache.kafka.clients.admin|Provides a Kafka client for performing administrative operations (such as creating topics and configuring brokers) on a Kafka cluster.
org.apache.kafka.clients.consumer|Provides a Kafka client for consuming records from topics and/or partitions in a Kafka cluster.
org.apache.kafka.clients.producer|Provides a Kafka client for producing records to topics and/or partitions in a Kafka cluster.
org.apache.kafka.common|Provides shared functionality for Kafka clients and servers.
org.apache.kafka.common.acl|Provides classes representing Access Control Lists for authorization of clients
org.apache.kafka.common.annotation|Provides annotations used on Kafka APIs.
org.apache.kafka.common.config|Provides common mechanisms for defining, parsing, validating, and documenting user-configurable parameters.
org.apache.kafka.common.config.provider|Provides a pluggable interface and some implementations for late-binding in configuration values.
org.apache.kafka.common.errors|Provides common exception classes.
org.apache.kafka.common.header|Provides API for application-defined metadata attached to Kafka records.
org.apache.kafka.common.metrics|Provides the API used by Kafka clients to emit metrics which are then exposed using the * MetricsReporter interface.
org.apache.kafka.common.metrics.stats|Provides methods of statistically aggregating metrics upon emission.
org.apache.kafka.common.quota|Provides mechanisms for enforcing resource quotas.
org.apache.kafka.common.resource|Provides client handles representing logical resources in a Kafka cluster.
org.apache.kafka.common.security.auth|Provides pluggable interfaces for implementing Kafka authentication mechanisms.
org.apache.kafka.common.security.oauthbearer|Provides a LoginModule for using OAuth Bearer Token authentication with Kafka clusters.
org.apache.kafka.common.security.oauthbearer.secured|This package is deprecated.
org.apache.kafka.common.security.plain|Provides implementation to use plaintext credentials authentication for securing Kafka clusters.
org.apache.kafka.common.security.scram|Provides adaptor to use the Salted Challenge Response Authentication Mechanism for securing Kafka clusters.
org.apache.kafka.common.security.token.delegation|Provides mechanism for delegating authorization to a distinct Principal for securing Kafka clusters.
org.apache.kafka.common.serialization|Provides interface and some implementations of serialization/deserialization routines for various objects.
org.apache.kafka.server.authorizer|Provides pluggable interface for performing authorization on a Kafka server.
org.apache.kafka.server.policy|Provides pluggable interfaces for expressing policies on topics and configs.
org.apache.kafka.server.quota|Provides pluggable interface for enforcing client quotas from a Kafka server.


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

gharris1727 committed:
KAFKA-13881: Add package descriptions for clients packages

Signed-off-by: Greg Harris <greg.harris@aiven.io>

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Kafka Client for performing administrative operations on a Kafka cluster
```
```suggestion
 * Provides a Kafka client for performing administrative operations (such as creating topics and configuring brokers) on a Kafka cluster.
```

tombentley commented on clients/src/main/java/org/apache/kafka/clients/consumer/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Kafka Client for consuming events from a Kafka Cluster
```
```suggestion
 * Provides a Kafka client for consuming records from topics and/or partitions in a Kafka cluster.
```

tombentley commented on clients/src/main/java/org/apache/kafka/clients/producer/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Kafka Client for producing events to a Kafka Cluster
```
```suggestion
 * Provides a Kafka client for producing records to topics and/or partitions in a Kafka cluster.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/acl/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Access Control List API for Authorization of Kafka Clients
```
```suggestion
 * Provides classes representing Access Control Lists for authorization of clients.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/annotation/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Annotations for describing properties of Kafka API Classes
```
```suggestion
 * Provides annotations used on Kafka APIs.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/cache/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Reusable implementations of cache primitives
```
```suggestion
 * Reusable implementations of cache primitives.
 * <strong>This package is not a supported Kafka API; the implementation may change without warning between minor or patch releases.</strong>
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/compress/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Mechanisms for compressing data handled by Kafka
```
```suggestion
 * Mechanisms for compressing data handled by Kafka.
 * <strong>This package is not a supported Kafka API; the implementation may change without warning between minor or patch releases.</strong>
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/config/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Standard mechanisms for defining, parsing, validating, and documenting user-configurable parameters
```
```suggestion
 * Provides common mechanisms for defining, parsing, validating, and documenting user-configurable parameters.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/config/provider/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Pluggable interface and implementations for late-binding in configuration values
```
```suggestion
 * Provides a pluggable interface and some implementations for late-binding in configuration values.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/config/types/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Custom non-primitive types of configuration properties
```
```suggestion
 * Provides custom non-primitive types of configuration properties.
 * <strong>This package is not a supported Kafka API; the implementation may change without warning between minor or patch releases.</strong>
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/errors/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Errors that are thrown by Kafka Clients
```
```suggestion
 * Provides common exception classes.
```

tombentley reviewed (commented):
Thanks! I have few initial points:

* While Kafka should start with a capital letter there is no need for "Clients" to.
* We should use complete sentences, finishing with a full stop.
* It would be great to call out those packages which are in the clients jar but are not publicly supported APIs. The publicly supported ones are those listed at https://kafka.apache.org/33/javadoc/index.html. This will prevent people accidentally finding and using them by mistake.
* The direct subpackages of `org.apache.kafka.common` are generally classes which are used in common by both clients and servers. Using text like "... on Kafka clients" for such packages is misleading. 

I left a few examples, to illustrate.

gharris1727 committed:
Review comments: grammar, capitalization, some internal package warnings.

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comments: add more internal package warnings

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 commented:
Thanks for the review round @tombentley, I've edited all of the package descriptions to more closely follow your recommendations. Let me know if any still miss the mark.

tombentley commented on clients/src/main/java/org/apache/kafka/common/header/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Provides API for application defined metadata attached to Kafka records.
```
```suggestion
 * Provides API for application-defined metadata attached to Kafka records.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/metrics/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Provides mechanisms for emitting monitoring metrics.
```
```suggestion
 * Provides the API used by Kafka clients to emit metrics which are then exposed using the {@link MetricsReporter} interface.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/network/package-info.java:
```diff
@@ -0,0 +1,21 @@
+ * limitations under the License.
+ */
+/**
+ * Provides mechanism for sending to and receiving data from remote machines.
```
```suggestion
 * Provides the network API used by the Kafka clients.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/record/package-info.java:
```diff
@@ -0,0 +1,21 @@
+ * limitations under the License.
+ */
+/**
+ * Provide classes representing a single record in a Kafka topic and/or partition.
```
```suggestion
 * Provides the low-level representation of records and record batches used by clients and servers.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/security/auth/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Provides pluggable interface for implementing Kafka authentication mechanisms.
```
```suggestion
 * Provides pluggable interfaces for implementing Kafka authentication mechanisms.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Provides adaptor for using OAuth Bearer Token Authentication for securing Kafka clusters.
```
```suggestion
 * Provides a {@code LoginModule} for using OAuth Bearer Token authentication with Kafka clusters.
```

tombentley commented on clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/package-info.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * This package is deprecated.
+ */
+package org.apache.kafka.common.security.oauthbearer.secured;
```
You could add the `@Deprecated` annotation and the `@deprecated` Javadoc tag.

gharris1727 replied:
I added the `@deprecated` Javadoc tag, but needed to leave the original `This package is deprecated.` otherwise the package listing was empty.

I also tried adding `@Deprecated` to the package name, and received a build error.
Let me know if the current state is still not acceptable.

tombentley commented on clients/src/main/java/org/apache/kafka/server/policy/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+/**
+ * Provides pluggable interface for authorization policies.
```
```suggestion
 * Provides pluggable interfaces for expressing policies on topics and configs.
```

mimaison reviewed (commented):
Thanks for the PR, this is a nice improvement. Including all suggestions from @tombentley, LGTM

gharris1727 committed:
Review comments: edits for clients package infos

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
KAFKA-13881: Add connect package infos

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
KAFKA-13881: Add Streams package infos

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 commented:
@mimaison Thanks for taking a look.

I've implemented the suggested changes, but also added the `connect` and `streams` package descriptions which need a pass. In particular I'm not familiar with streams, and the descriptions are my best guess based on perusing the public documentation.

Thanks!

mimaison commented:
Thanks for the update! I wonder if it would be better to have Connect and Streams in separate PRs especially as it's likely going to be different people reviewing them.

gharris1727 committed:
Revert "KAFKA-13881: Add connect package infos"

This reverts commit 4cc2cd9af0ad221d5f22340cdbeeb246689e0d3f.

gharris1727 committed:
Revert "KAFKA-13881: Add Streams package infos"

This reverts commit 164b2ebe6fdb9fade463960aa2c26ddc485b8732.

gharris1727 commented:
@mimaison I had some wishful thinking that putting everything together would mean that we're more motivated to get everything in, but I agree that the packages are logically separate enough to require separate review. I've opened the split PRs and removed connect and streams from this PR.

mimaison reviewed (approved):
LGTM

mimaison commented:
@gharris1727 Can you rebase this PR? It looks like the commit it's currently on is messing with the CI. Thanks

gharris1727 committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-13881-package-info

mjsax commented:
Cherry-picked to `3.4` branch.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/replica/package-info.java:
```diff
@@ -0,0 +1,21 @@
+ * limitations under the License.
+ */
+/**
+ * Provides mechanism for determining placement of Kafka log replicas.
```
I'm curious why the `ReplicaSelector` plugin is NOT considered a public API?
